### ISSUE TITLE
Setting ep_integrate to true means no results

### ISSUE THREAD

**Issue by Pezzab** [10up/ElasticPress#892]
I have a custom post with additional custom fields and have created a custom WP_Query in my search.php.

If I set` ep_integrate` to `true` I get no results, if to` false` I get all the results I expect.

I'm aware I'm at the bottom of a steep learning curve re Elasticsearch and Elasticpress so excuse my ignorance but does setting` ep_integrate` to` false` mean I'm back to searching the MYSQL DB directly again (obviously no good)?

If so how do I get Elasticpress/search to play nice with my query?

**Comment by ivankristianto** [comment]
@Pezzab yes if you set `ep_integrate` to false, it will query to mysql.
have try to reindex your site?

**Comment by Pezzab** [comment]
The site is indexed OK.

If I set it to `true` and run the search there are no results at the front end but looking in Debug I end up with 2 Elasticpress queries. One has all the expected results but isn't displayed whilst the second has none.

Setting it back to `false` again I only get 1 Elasticpress query in Debug (the one with the results) though I assume this is not where the data is being pulled from when the search results are shown at the front end.

**Comment by Pezzab** [comment]
If I set `ep_integrate` to `true` why do I get 2 queries and why does it not display the successful query instead of the empty one?

That said even with it set to` false` it seems that Elasticpress is providing the results - I say this because I have a filter in `functions.php` to switch off fuzziness and if I turn it off or on this appears to affect the number of results returned. In fact they match the result shown in Debug.

**Comment by Pezzab** [comment]
OK I've just realised something odd. While 'false' the search is definitely only returning results via MYSQL but even so the number of results pages is being governed by the Elasticsearch result.

My search form has an option to filter by a custom taxonomy. Regardless of how frequent the occurrence of a given search term under a given taxonomy the number of page results remains the same.

It's a museum collections search I'm working on so if I search for "sherd" in Egyptology I expect a lot of results. Searching the same term in Social history I only get 1 result. Even so both results return 51 pages, the second one has the same single result spread over those 51 pages.

So it seems Elasticpress is influencing the display of results even if the results aren't coming via Elasticsearch itself, if that makes sense.

**Comment by ivankristianto** [comment]
@Pezzab 
This is a strange issue indeed.
I was wondering why the ES query twice in your case.
Can you install debug bar elasticpress and send me both of the ES query body to my email ( ivan.kristianto[at]10up.com ) or attached here also fine.
To install the debug bar please follow this article: https://www.elasticpress.io/blog/2017/05/debugging/

**Comment by Pezzab** [comment]
Thanks Ivan,

I've sent you the attachment.

**Comment by ivankristianto** [comment]
@Pezzab I did take a look at your log.
You have wrong way to do the search.
1. When ElasticPress is active, when you have search params, it will automatically add `ep_integrate` by default.
2. Looks like you are in search page, then you do another WP_Query because you want to add `tax_query`, thus the search will be done twice. First is from WordPress search default, second is from you custom WP_Query.

So if you want to add tax_query to your search query. I suggest you to use filter `pre_get_posts` instead create new WP_Query.
Codex: https://codex.wordpress.org/Plugin_API/Action_Reference/pre_get_posts

**Comment by Pezzab** [comment]
Thank you Ivan,

That makes much more sense now

**Comment by Pezzab** [comment]
I still can't seem to get any results even using `pre_get_posts`:
```

    $taxquery = array(
        array(
            'taxonomy' => 'department',
            'field' => 'slug',
            'terms' => $dept,
            'operator'=> 'IN'
        )
    );

    $query->set( 'tax_query', $taxquery );
```
This is what comes up in ElasticPress Query Body - the taxonomy I want to filter by appears but the search yields no result (I know there should be at least one result):
```

{
    "from": 0,
    "size": 9,
    "sort": [
        {
            "_score": {
                "order": "desc"
            }
        }
    ],
    "query": {
        "function_score": {
            "query": {
                "bool": {
                    "should": [
                        {
                            "multi_match": {
                                "query": "onions",
                                "type": "phrase",
                                "fields": [
                                    "post_title",
                                    "post_content",
                                    "post_excerpt",
                                    "author_name",
                                    "terms.post_tag.name",
                                    "terms.category.name",
                                    "post_author.login"
                                ],
                                "boost": 4
                            }
                        },
                        {
                            "multi_match": {
                                "query": "onions",
                                "fields": [
                                    "post_title",
                                    "post_content",
                                    "post_excerpt",
                                    "author_name",
                                    "terms.post_tag.name",
                                    "terms.category.name",
                                    "post_author.login"
                                ],
                                "boost": 2,
                                "fuzziness": 0,
                                "operator": "and"
                            }
                        },
                        {
                            "multi_match": {
                                "fields": [
                                    "post_title",
                                    "post_content",
                                    "post_excerpt",
                                    "author_name",
                                    "terms.post_tag.name",
                                    "terms.category.name",
                                    "post_author.login"
                                ],
                                "query": "onions",
                                "fuzziness": 0
                            }
                        }
                    ]
                }
            },
            "exp": {
                "post_date_gmt": {
                    "scale": "14d",
                    "decay": 0.25,
                    "offset": "7d"
                }
            },
            "score_mode": "avg",
            "boost_mode": "sum"
        }
    },
    "post_filter": {
        "bool": {
            "must": [
                {
                    "bool": {
                        "must": [
                            {
                                "terms": {
                                    "terms.department.slug": [
                                        "Fine art"
                                    ]
                                }
                            }
                        ]
                    }
                },
                {
                    "term": {
                        "post_type.raw": "object"
                    }
                },
                {
                    "term": {
                        "post_status": "publish"
                    }
                }
            ]
        }
    }
}
```

**Comment by Pezzab** [comment]
Just to be clear the above code works if I deactivate ElasticPress.

**Comment by ivankristianto** [comment]
I see your tax_query looking for field `slug` but you give name instead.
So your tax_query should be:
```
$taxquery = array(
        array(
            'taxonomy' => 'department',
            'field' => 'name',
            'terms' => $dept,
            'operator'=> 'IN'
        )
    );

    $query->set( 'tax_query', $taxquery );
```

**Comment by Pezzab** [comment]
Sorry Ivan,
I managed to paste the wrong version of my code (while trying different alternatives), I was already using "name".

I still get the same issue.

**Comment by ivankristianto** [comment]
hmm.. That's weird though.
when you search this "query": "onions", is the keyword in one of the search fields?
can you resync the es index and try again?

**Comment by Pezzab** [comment]
Yes it's in the title field (It's a random example but I use it because I know there is only one occurrence in the data - it's a title for a painting).

I've already tried re-indexing.

If I reactivate ElasticPress but _disable_ `my pre_get_posts` function I get 78 results (but obviously can no longer filter by taxonomy), my most relevant, unique result comes top then the next one is a fuzzy match for "Union".

So ElasicPress is working it just doesn't like my `pre_get_posts` function which seems OK and is otherwise pretty straightforward (and works when I disable EP). But I really need to be able to filter search results to make the project I'm working on viable (200,000 object records so WordPress native search isn't going to cut it).

I've also tried disabling other plugins to rule that out btw.

**Comment by Pezzab** [comment]
OK I think I've worked out what the problem is.

EP is working fine, the issue was my not correctly matching the case of the department name with the query slug: e.g. Fine Art != fine art or fine-art etc.

Changing the string case of the query slug in `pre_get_posts` to match the taxonomy name fixed the error.

ElasticSearch appears to be working fine.

**Comment by ivankristianto** [comment]
Glad to hear it's working. There is an iaaue with case sensitive search in ES 5.1 and get fixed in 5.2.
The next release of EP will support that

**Comment by Pezzab** [comment]
Thanks for your patience btw Ivan!
